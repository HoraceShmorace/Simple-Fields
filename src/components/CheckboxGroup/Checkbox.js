import { useState } from 'react'
import { bool, func, oneOf, string } from 'prop-types'
import cx from 'classnames'
import styles from './Checkbox.module.css'

const Checkbox = ({ description, id, isDefaultChecked, labelFirst, label, onClick, type, value }) => {
  const [isChecked, setIsChecked] = useState(isDefaultChecked)
  const className = cx([
    styles.root,
    styles[type],
    { [styles.checked] : isChecked},
    labelFirst ? styles['label-before'] : styles['label-after']
  ])
  
  const toggleClick = () => {
    const nextIsChecked = !isChecked
    setIsChecked(nextIsChecked)
    if (typeof onClick === 'function') onClick(nextIsChecked)
  }

  return (
    <li className={className}>
      <input id={id} type="checkbox" value={value} defaultChecked={isChecked} onClick={toggleClick} />
      {
        labelFirst && 
        <label htmlFor={id}>
          {label}
        </label>
      }
      <button onClick={toggleClick}>
        <div className={styles.indicator} />
      </button>
      {
        !labelFirst && 
        <label htmlFor={id}>
          {label}
        </label>
      }
    </li>
  )
}

Checkbox.propTypes = {
  description: string,
  id: string.isRequired,
  isDefaultChecked: bool,
  labelFirst: bool,
  label: string,
  onClick: func,
  type: oneOf([
    'checkbox',
    'switch',
    'pills'
  ]),
  value: string
}

Checkbox.defaultProps = {
  description: null,
  isDefaultChecked: false,
  labelFirst: false,
  label: null,
  onClick: null,
  type: 'checkbox',
  value: ''
}

export default Checkbox
