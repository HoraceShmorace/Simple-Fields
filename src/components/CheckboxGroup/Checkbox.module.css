.root {
  align-items: center;
  display: flex;
  flex-direction: row;
  margin-bottom: 1rem;
}

.label-after label {
  margin-left: 0.75rem;
}

.label-before label {
  margin-right: 0.75rem;
}

.root input[type=checkbox] {
  visibility: hidden;
  position: absolute;
}

.root label {
  color: #666;
  cursor: pointer;
  font-size: 1rem;
  font-weight: bold;
}

.checkbox button {
  background-color: transparent;
  border: 3px solid #999;
  border-radius: 0.4rem;
  cursor: pointer;
  height: 1.5rem;
  padding: 3px;
  position: relative;
  width: 1.5rem;
}

.checkbox button div.indicator {
  background-color: #ccc;
  border-radius: 0.2rem;
  cursor: pointer;
  height: 90%;
  margin: auto;
  width: 90%;
}

.checkbox.checked button {
  background-color: #fff;
  border-color: #444;
}

.checkbox.checked button div.indicator {
  background-color: #11ccff;
  height: 99%;
  width: 99%;
}

.switch button {
  background-color: #bbb;
  border: 2px solid #ccc;
  border-radius: 1rem;
  cursor: pointer;
  display: flex;
  height: 1rem;
  padding: 0;
  width: calc(2rem + 4px);
}

.switch .indicator {
  background-color: #aaa;
  border: 3px solid white;
  border-radius: 1rem;
  height: 1rem;
  margin-top: -5px;
  transform: translateX(-2px);
  transition: all 300ms;
  width: 1rem;
}

.switch.checked button div.indicator {
  background-color: #11aa11;
  transform: translateX(calc(1rem - 4px));
}
