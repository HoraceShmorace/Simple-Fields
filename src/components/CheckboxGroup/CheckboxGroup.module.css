.root ul {  
  list-style-type: none;
  margin: 0;
  padding-inline-start: 0;
  padding-left: 0;
}

.root ul li:last-child {
  margin-bottom: 0;
}

.pill ul {
  margin: -0.2rem 0 0 -0.2rem;
}
