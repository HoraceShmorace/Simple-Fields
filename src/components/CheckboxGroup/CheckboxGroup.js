import Fieldset from '../Fieldset/Fieldset'
import Checkbox from './Checkbox'
import styles from './CheckboxGroup.module.css'

const CheckboxGroup = ({ fields, label }) => (
  <Fieldset className={styles.root} label={label}>
    <ul>
    { fields.map(props => <Checkbox key={props.id} {...props} />)}
    </ul>
  </Fieldset>
)

export default CheckboxGroup
