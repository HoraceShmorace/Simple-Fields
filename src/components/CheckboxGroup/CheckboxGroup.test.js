import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import CheckboxGroup from './CheckboxGroup'
import mockData from '~mock/form.json'

const FIELD_TYPE_TEXT = 'checkbox'

describe('CheckboxGroup', () => {
  describe('when rendered', () => {
    const props = mockData.fields.find(({ type }) => type === FIELD_TYPE_TEXT)
    const component = shallow(
      <CheckboxGroup {...props} />
    )

    it('should match snapshot', () => {
      expect(toJson(component)).toMatchSnapshot()
    })
  })
})
