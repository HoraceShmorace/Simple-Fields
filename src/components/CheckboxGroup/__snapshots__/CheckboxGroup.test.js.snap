// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup when rendered should match snapshot 1`] = `
<ul
  className="fs-checkbox-group"
>
  <li
    key="Option1"
  >
    <Checkbox
      id="115390353_1"
      isChecked={false}
      label="Option1"
      onClick={[Function]}
      value="Option1"
    />
  </li>
  <li
    key="Option2"
  >
    <Checkbox
      id="115390353_2"
      isChecked={false}
      label="Option2"
      onClick={[Function]}
      value="Option2"
    />
  </li>
  <li
    key="Option3"
  >
    <Checkbox
      id="115390353_3"
      isChecked={false}
      label="Option3"
      onClick={[Function]}
      value="Option3"
    />
  </li>
</ul>
`;
